quinn-proto/src/connection/assembler.rs

```diff
@@ -64,68 +67,48 @@ impl Assembler {
             }
 
             return Some(if max_length < chunk.bytes.len() {
                 self.bytes_read += max_length as u64;
                 let offset = chunk.offset;
                 chunk.offset += max_length as u64;
+                self.buffered -= max_length;
                 Chunk::new(offset, chunk.bytes.split_to(max_length))
             } else {
                 self.bytes_read += chunk.bytes.len() as u64;
-                self.defragmented = self.defragmented.saturating_sub(1);
+                self.buffered -= chunk.bytes.len();
+                self.allocated -= chunk.allocation_size;
                 let chunk = PeekMut::pop(chunk);
                 Chunk::new(chunk.offset, chunk.bytes)
             });
         }
     }
 
-    // Copy the buffered chunk data to new chunks backed by a single buffer to
-    // make sure we're not unnecessarily holding on to many larger allocations.
-    // Merge contiguous chunks in the process of doing so. Reset the `defragmented`
-    // counter to the new number of chunks left in the heap so that we can decide
-    // when to defragment the queue again if necessary.
-    fn defragment(&mut self) {
-        let buffered = self.data.iter().map(|c| c.bytes.len()).sum::<usize>();
-        let mut buffer = BytesMut::with_capacity(buffered);
-        let mut offset = self
-            .data
-            .peek()
-            .as_ref()
-            .expect("defragment is only called when data is buffered")
-            .offset;
-
-        let new = BinaryHeap::with_capacity(self.data.len());
-        let old = mem::replace(&mut self.data, new);
-        for chunk in old.into_sorted_vec().into_iter().rev() {
-            let end = offset + (buffer.len() as u64);
-            if let Some(overlap) = end.checked_sub(chunk.offset) {
-                if let Some(bytes) = chunk.bytes.get(overlap as usize..) {
-                    buffer.extend_from_slice(bytes);
-                }
-            } else {
-                let bytes = buffer.split().freeze();
-                self.data.push(Buffer { offset, bytes });
-                offset = chunk.offset;
-                buffer.extend_from_slice(&chunk.bytes);
-            }
-        }
-
-        let bytes = buffer.split().freeze();
-        self.data.push(Buffer { offset, bytes });
-        self.defragmented = self.data.len();
-    }
-
-    pub(crate) fn insert(&mut self, mut offset: u64, mut bytes: Bytes) {
+    // Note: If a packet contains many frames from the same stream, the estimated over-allocation
+    // will be much higher because we are counting the same allocation multiple times.
+    pub(crate) fn insert(&mut self, mut offset: u64, mut bytes: Bytes, allocation_size: usize) {
+        debug_assert!(
+            bytes.len() <= allocation_size,
+            "allocation_size less than bytes.len(): {:?} < {:?}",
+            allocation_size,
+            bytes.len()
+        );
+        self.end = max(self.end, offset + bytes.len() as u64);
```

Comment:
Nix: the style in most other files is `self.end = self.end.max(offset + bytes.len() as u64);`
This avoids the need for the import

Reply:
Fixed.